Backpack/src/main/java/net/skyscanner/backpack/contentbubble/BpkContentBubble.kt

```diff
@@ -0,0 +1,224 @@
+      _round = value
+      requestLayout()
+    }
+
+  private var _pointerPosition = PointerPosition.MIDDLE
+  /**
+   * Specify where the pointer should be rendered.
+   *
+   * @see [PointerPosition]
+   */
+  var pointerPosition: PointerPosition
+    get() = _pointerPosition
+    set(value) {
+      _pointerPosition = value
+      requestLayout()
+    }
+
+  init {
+    this.initialize(attrs, defStyleAttr)
+  }
+
+  private fun initialize(attrs: AttributeSet?, defStyleAttr: Int) {
+    context.theme.obtainStyledAttributes(attrs, R.styleable.BpkContentBubble, defStyleAttr, 0)
+      ?.use {
+        _fitContent = it.getBoolean(R.styleable.BpkContentBubble_contentBubbleFitContent, fitContent)
+        _round = it.getBoolean(R.styleable.BpkContentBubble_contentBubbleRound, round)
+
+        _pointerPosition = it.getInt(R.styleable.BpkContentBubble_contentBubblePointerPosition, pointerPosition.id)
+          .let(::mapXmlToPointerPosition) ?: pointerPosition
+      }
+
+    background = null
+    setWillNotDraw(false)
+    setPadding(0, 0, 0, 0)
+    // PorterDuffXfermode doesn't work in some devices when hardware acceleration is on
+    setLayerType(LAYER_TYPE_SOFTWARE, null)
+
+    pointerDrawable.apply {
+      setBounds(0, 0, intrinsicWidth, intrinsicHeight)
+    }
+
+    cornerRadiusDrawable.apply {
+      setBounds(0, 0, intrinsicWidth, intrinsicHeight)
+    }
+
+    setUpMasks()
+  }
+
+  override fun draw(canvas: Canvas) {
+    val count = canvas.saveCount
+
+    val width = width.toFloat()
+    val height = height.toFloat()
+
+    val pointerHalfWidth = pointerDrawable.intrinsicWidth / 2
+    val pointerOffset = if (layoutDirection == View.LAYOUT_DIRECTION_RTL) {
+      1 - pointerPosition.offset
+    } else {
+      pointerPosition.offset
+    }
+
+    val yStart = height - pointerDrawable.intrinsicHeight
+    val pointerXStart = width * pointerOffset - pointerHalfWidth
+    val pointerXEnd = width * pointerOffset + pointerHalfWidth
+
+    clipRect.set(0f, yStart, pointerXStart, height)
+    canvas.clipOutRectCompat(clipRect)
+
+    clipRect.set(width, yStart, pointerXEnd, height)
+    canvas.clipOutRectCompat(clipRect)
+
+    super.draw(canvas)
+    canvas.restoreToCount(count)
+
+    paint.xfermode = porterDuffXfermode
+    canvas.drawBitmap(mask, pointerXStart, yStart, paint)
+
+    if (round) {
+      drawRadiusMask(canvas)
+    }
+
+    paint.xfermode = null
+  }
+
+  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+    val firstChild = getChildAt(0)
+    if (firstChild == null || !fitContent || firstChild.layoutParams.height >= 0) {
+      super.onMeasure(widthMeasureSpec, heightMeasureSpec)
+    } else {
+      getChildAt(0)?.let {
+        // When fit content is true we add the size of the pointer to the view's height, in order
+        // for the children to take up that size we need to change `WRAP_CONTENT` to `MATCH_PARENT`
+        if (it.layoutParams.height == ViewGroup.LayoutParams.WRAP_CONTENT) {
```

Comment:
It's usually a bad practice to modify LayoutParams in ViewGroup. Let's use padding modification instead

Reply:
removed